### PR TITLE
[WIP] publish on google cloud run

### PR DESCRIPTION
**Pull request by rprimet**
This is a very rough draft to start a discussion on a possible datasette cloud run publish plugin (see issue #400).

The main change was to dynamically set the listening port in `make_dockerfile` to satisfy cloud run's [requirements](https://cloud.google.com/run/docs/reference/container-contract).

This was done by running `datasette` through `sh` to get environment variable substitution. Not sure if that's the right approach?

### CODE REVIEW

**Comment by simonw** [comment]
Thanks for looking into this!

To clarify: currently, the Dockerfile that we generate looks something like this:

```
CMD ["datasette", "serve", "--host", "0.0.0.0", "fixtures.db", "--cors", "--port", "8001"]
```

Your code here changes that CMD line to look like this instead, in order to set the port based on an environment variable:

```
CMD ["sh", "-c", "datasette serve --port $PORT ..."]
```

I wonder if this is the only way to do this?

File: datasette/publish/cloudrun.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+                "title": title,
+                "license": license,
+                "license_url": license_url,
+                "source": source,
+                "source_url": source_url,
+                "about": about,
+                "about_url": about_url,
+            },
+        ):
+            image_id = f"gcr.io/{project}/{name}"
```
Comment (simonw) [inline datasette/publish/cloudrun.py]: Datasette still supports Python 3.5 so we can't use format strings just yet.

**Comment by simonw** [comment]
I asked @andrewgodwin about this and he confirmed that if we want to read an environment variable we can't use the `CMD [...]` syntax in the way that we were using it.

He did suggest that if we're doing `CMD ["sh", "-c", "datasette serve --port $PORT ..."]` we may as well do this instead:

`CMD "datasette serve --port $PORT ..."`

We should apply some command-line escaping here - if the user passes `--version-note=hello$there` to `datasette publish` we need that $ not to be accidentally evaluated as an environment variable.

It looks like [shlex.quote](https://docs.python.org/dev/library/shlex.html#shlex.quote) is the right way to do that.

**Comment by simonw** [comment]
This is amazing - works an absolute treat. Thank you very much!

**Comment by eyeseast** [comment]
This is exactly what I needed. Thank you.

**Comment by simonw** [comment]
Here's my first working deployment: https://datasette-j7hipcg4aq-uc.a.run.app/fixtures-c35b6a5/facetable?_facet_array=tags

I deployed it using this:

    datasette publish cloudrun fixtures.db --branch=master

The second time I ran the command I got an error:

    ERROR: (gcloud.beta.run.deploy) Deployment endpoint was not found. Perhaps the
    provided region was invalid. Set the `run/region` property to a valid region and
    retry. Ex: `gcloud config set run/region us-central1`

So I ran the command it suggested and then everything worked:

    gcloud config set run/region us-central1
    datasette publish cloudrun fixtures.db --branch=master

**Comment by simonw** [comment]
Documentation is now available here: https://datasette.readthedocs.io/en/latest/publish.html#publishing-to-google-cloud-run

**Comment by rprimet** [comment]
> The second time I ran the command I got an error:

ERROR: (gcloud.beta.run.deploy) Deployment endpoint was not found. Perhaps the
provided region was invalid. Set the `run/region` property to a valid region and
retry. Ex: `gcloud config set run/region us-central1`

Yes, I was able to reproduce this; I used to get prompted for a run region interactively by the `gcloud` tool before, but maybe this is changing? (the [documentation](https://cloud.google.com/run/docs/deploying) now assumes `run/region` is set).

Not sure which course of action is best: making `datasette` ensure that `run/region` is set beforehand or wait a bit until the gcloud CLI stabilizes?

**Comment by simonw** [comment]
Since there's a useful error message I'm OK with revisiting this in a few weeks to see if they change the CLI tool.

**Pull request merged**